File: src/data/algorithms/bubble.ts
import { delay } from "../../utils/delay";

// ! bubble sort
export const bubbleSort = async (
  nums: number[],
  setArray: React.Dispatch<React.SetStateAction<number[]>>,
  setSelectedIndices: React.Dispatch<React.SetStateAction<number[]>>,
  ms: number
): Promise<void> => {
  for (let i = 0; i < nums.length - 1; i++) {
    for (let j = 0; j < nums.length - i - 1; j++) {
      setSelectedIndices([j, j + 1]);
      if (nums[j] > nums[j + 1]) {
        [nums[j], nums[j + 1]] = [nums[j + 1], nums[j]];
      }
      await delay(ms);
      setArray(nums);
    }
  }

  setSelectedIndices([]);

  for (let i = 0; i < nums.length; i++) {
    setSelectedIndices((value) => [...value, i]);
    await delay(ms);
  }

  await delay(ms);
  setSelectedIndices([]);
};

File: code/bubble/bubble.go
package bubble

func bubbleSort(nums []int) []int {
	N := len(nums)
	for i := 0; i < N; i++ {
		for j := 0; j < N-i-1; j++ {
			if nums[j] > nums[j+1] {
				nums[j], nums[j+1] = nums[j+1], nums[j]
			}
		}
	}
	return nums
}
